### ISSUE TITLE
Add support for event deletion on snapshot predicate

### ISSUE THREAD

**Issue by octonato** [akka/akka#29685]
Currently, it's only possible to delete events when using a snapshot strategy based on a counter and retention criteria. The predicated variant doesn't offer this option. 

Being able to save a snapshot and deleted previous events with a predicate is particularly useful. One can implement tombstones:

> when that event happens, transition to that state, save a snapshot and delete previous events

This is partially triggered by this other issue: #29684, but I believe that it has value in itself. Even if we realise that it can't be used in the context of #29684.

**Comment by fluxlife** [comment]
I would like to second this request. Not only is this important for custom tombstoning and _potentially_ other performance considerations, but would be very helpful from a GDPR perspective.

**Comment by dompetrauskas** [comment]
Is anyone working on this? As far as I know, from my familiarity with `EventSourcedBehavior` implementation, it shouldn't be too complicated. I could give it a shot.

**Comment by johanandren** [comment]
I don't think anyone is working on this, please go ahead!

**Comment by ptrdom** [comment]
I wonder if this functionality will cause any issues with out of sync read-sides, particularly in the case this is used to implement tombstones described in https://github.com/akka/akka-persistence-jdbc/issues/246. By that I mean that in Cassandra plugin, event deletion does not delete tag views that are kept in a separate table, so the downside is that the events remain, but read-sides can reliably process the event that caused deletion. In case of JDBC plugin, all data is stored in a single table, so there is a potential outcome where read-sides do not get to process the event that caused deletion before it gets deleted, in turn orphaning/corrupting whatever data was projected by read-sides. It would be neat to implement retention to be configurable so it supports intended use case expressed by it, but maybe this is becoming to opinionated/complicated and the developers using either of plugins just got to be aware of this behavior and that is it. What do you think?

**Comment by fluxlife** [comment]
Hmm, overall it seems what's challenging is that there is no concensus between Projection plugins and Persistence plugins on how to handle deleted events. For example, it's not clear to me, in the situation where events are soft-deleted, do projections that are behind in events still project the soft-deleted events? The current projection API via the EventEnvelope I believe doesn't let us know if the event that is being consumed is soft-deleted or not (i.e. there is no `eventEnvelope.isDeleted` etc)

I dont know how realistic it is but it almost seems like if it were possible to separate concerns of marking something as deleted vs actually deleting from persistence, that would be ideal. Then that would give users complete control of how to deal with events 'marked for deletion' and let them handle when to hard-delete them from their persistence layer.

**Comment by ptrdom** [comment]
There are probably multiple ways to go about it with their own pros and cons. I was thinking about a projection that would emit both events and snapshots in orderly fashion.

**Comment by ptrdom** [comment]
Also another matter is that `EventSourcedBehavior` does not wrap saving of snapshot in a same transaction that processes command and persist events - no matter if snapshot saving fails or succeeds, behavior continues consuming commands, and there is no automated reconciliation. The functionality that we are trying to express here with snapshot saving and event retention requires stronger guarantees than those actions currently provide.

**Comment by ptrdom** [comment]
Maybe this warrants adding specialized effects to `EventSourcedBehavior`, like `Effect.snapshot` that would make sure to run persisting of events and saving of snapshot in a single transaction, with all associated costs of that.

**Comment by johanandren** [comment]
Some backends won't even have that capability anyway, and to complicate things even more the snapshots could even be saved in a completely different store than the events. Better to not encode something like looks like guarantees for that, which will only work in some use cases, and instead recommend that you keep 2+ snapshots and then delete events up to that (I think this is what is encoded in the auto-snapshot-delete-event logic).

**Comment by patriknw** [comment]
> overall it seems what's challenging is that there is no concensus between Projection plugins and Persistence plugins on how to handle deleted events

That's right. Deletes with the Cassandra journal will not delete the tagged events. For JDBC they are deleted immediately also for eventsByTag queries.

**Comment by jturnergresham** [comment]
I would also like to second this request - I have discussed a little here [in forums](https://discuss.lightbend.com/t/how-to-retire-aggregates-and-archive-events/9820).

In any high throughput usage of event sourcing over a growing population of entities a means for cleanly retiring entities into a longer term history (or offline) is likely to be required, and this seems to be key to achieving the tidy-up aspect of that.

**Comment by leviramsey** [comment]
It is possible to MacGyver this functionality (modulo what event deletion actually entails in your persistence plugin) with a `Tombstoned` event and a `ForceSnapshot` command/`SnapshotForced` event pair.  For an entity whose retention criteria is "keep _n_ snapshots and delete earlier events":

* `Effect.persist(Tombstoned)`
* project the `Tombstoned` event to an actor which sends `ForceSnapshot` commands and watches for `SnapshotForced` events until it has seen _n_ `SnapshotForced`s with no intervening events (an events-by-persistence-id query is useful for this)
* `SnapshotForced` events (not surprisingly) force a snapshot and don't otherwise affect state

Depending on the snapshot store, the older duplicate snapshots can be deleted afterwards (e.g. with the Cassandra `Cleanup` tool).

**Comment by jturnergresham** [comment]
Thanks - that is roughly the approach I was thinking we would need to take for now, although neater than mine, will look into.

**Comment by leviramsey** [comment]
Actually, that approach won't quite work, but an elaboration of that will: force a tombstone check before building the `EventSourcedBehavior` and if that comes back "tombstoned", then the `EventSourcedBehavior` gets a `RetentionCriteria.snapshotEvery(1, 1).withDeleteEventsOnSnapshot` (the "maybe DurableState if the persistence plugin cooperates" criteria) with a message to self which guarantees that it will persist a final no-op event to force a snapshot which deletes events.

**Comment by tomb50** [comment]
+1 for this question - we recently looked at triggering the deletion of events from the snapshot Predicate rather than the RetentitionCritiera. I will have a think on the approach outlined above @leviramsey , thanks.

Quick question, I am not actually seeing my events get deleted even when configured by the Retention Strategy:

` @Override
  public RetentionCriteria retentionCriteria() {
    return RetentionCriteria.snapshotEvery(1,1).withDeleteEventsOnSnapshot();
  }`
  
With the above, I am getting snapshots but events are not being deleted, I tried breakpoint my plugins' SnapshotStore.doDeleteAsync() implementation methods, but neither get hit. Is there a good place to breakpoint from the Akka side to see why deletion is not being attempted?

Thanks!

**Comment by leviramsey** [comment]
`SnapshotStore.doDeleteAsync` wouldn't necessarily be hit when deleting events.  That would be the journal plugin's `asyncDeleteMessagesTo` method.  Snapshot deletion in `withDeleteEventsOnSnapshot` only happens if events are successfully deleted.
